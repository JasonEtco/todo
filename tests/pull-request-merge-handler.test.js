const nock = require('nock')
const { Probot } = require('probot')

const { loadConfig, loadDiff } = require('./helpers')
const pullRequestClosed = require('./fixtures/payloads/pull_request.closed.json')
const plugin = require('..')

describe('pull-request-merged-handler', () => {
  let probot
  const event = { name: 'pull_request', payload: pullRequestClosed }

  beforeEach(() => {
    probot = new Probot({
      id: 1,
      githubToken: 'secret',
      throttleOptions: {
        enabled: false
      }
    })
    probot.load(plugin)
  })

  it('does nothing on an unmerged, closed PR', async () => {
    await probot.receive({
      ...event,
      payload: {
        ...event.payload,
        pull_request: { merged: false }
      }
    })
  })

  it('creates an issue', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates an issue with a truncated title', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('long-title')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates an issue without assigning anyone', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(200, {
        content: loadConfig('autoAssignFalse').toString('base64')
      })

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates an issue and assigns the configured user', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(200, {
        content: loadConfig('autoAssignString').toString('base64')
      })

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates an issue and assigns the configured users', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(200, {
        content: loadConfig('autoAssignArr').toString('base64')
      })

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('does not create any issues if no todos are found', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('none')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('does not create an issue that already exists', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: [{ title: 'I am an example title', state: 'open' }]
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates many (5) issues', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .times(5)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('many')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .times(5)
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('ignores changes to the config file', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('config')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('ignores changes to the bin directory', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('bin')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(200, {
        content: loadConfig('excludeBin').toString('base64')
      })

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('creates an issue with a body line', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: []
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('body')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

      .post('/repos/JasonEtco/tests/issues', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it.only('reopens a closed issue', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: [{ number: 1, title: 'I am an example title', state: 'closed' }]
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(404, {})
      .get('/repos/JasonEtco/.github/contents/.github/config.yml')
      .reply(404, {})

      .patch('/repos/JasonEtco/tests/issues/1', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

      .post('/repos/JasonEtco/tests/issues/1/comments', parameters => {
        expect(parameters).toMatchSnapshot()

        return true
      })
      .reply(201, {})

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })

  it('respects the reopenClosed config', async () => {
    const mock = nock('https://api.github.com')

      .head('/repos/JasonEtco/tests/pulls/21')
      .reply(200)

      .get('/search/issues')
      .query(true)
      .reply(200, {
        items: [{ title: 'I am an example title', state: 'closed' }]
      })

      .get('/repos/JasonEtco/tests/pulls/21')
      .matchHeader('accept', 'application/vnd.github.diff')
      .reply(200, (await loadDiff('basic')).data)

      .get('/repos/JasonEtco/tests/contents/.github/config.yml')
      .reply(200, {
        content: loadConfig('reopenClosedFalse').toString('base64')
      })

    await probot.receive(event)
    expect(mock.activeMocks()).toStrictEqual([])
  })
})
