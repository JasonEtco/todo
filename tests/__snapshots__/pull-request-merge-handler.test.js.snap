// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pull-request-merged-handler creates an issue 1`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "I am an example title",
}
`;

exports[`pull-request-merged-handler creates an issue and assigns the configured user 1`] = `
Object {
  "assignee": "matchai",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @matchai",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "I am an example title",
}
`;

exports[`pull-request-merged-handler creates an issue and assigns the configured users 1`] = `
Object {
  "assignees": Array [
    "JasonEtco",
    "matchai",
    "defunkt",
  ],
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco, @matchai and @defunkt",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "I am an example title",
}
`;

exports[`pull-request-merged-handler creates an issue with a body line 1`] = `
Object {
  "assignee": "JasonEtco",
  "body": "Don&#x27;t forget about this body text!

---

https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "I am an example title",
}
`;

exports[`pull-request-merged-handler creates an issue with a truncated title 1`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "This is a really, really, really, really, really, really, really, really, really...",
}
`;

exports[`pull-request-merged-handler creates an issue without assigning anyone 1`] = `
Object {
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L10-L15

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "I am an example title",
}
`;

exports[`pull-request-merged-handler creates many (5) issues 1`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/file.txt#L4

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "Title number 1",
}
`;

exports[`pull-request-merged-handler creates many (5) issues 2`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/file.txt#L4

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "Title number 2",
}
`;

exports[`pull-request-merged-handler creates many (5) issues 3`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/file.txt#L4

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "Title number 3",
}
`;

exports[`pull-request-merged-handler creates many (5) issues 4`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/file.txt#L4

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "Title number 4",
}
`;

exports[`pull-request-merged-handler creates many (5) issues 5`] = `
Object {
  "assignee": "JasonEtco",
  "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/file.txt#L4

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @JasonEtco.",
  "labels": Array [
    "todo :spiral_notepad:",
  ],
  "title": "Title number 5",
}
`;

exports[`pull-request-merged-handler reopens a closed issue 1`] = `
Object {
  "state": "open",
}
`;

exports[`pull-request-merged-handler reopens a closed issue 2`] = `
Object {
  "body": "This issue has been reopened because the **\`TODO\`** comment still exists in [**.travis.yml**](https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml), as of 6c2373ad036d292f7e9467981ba8b467e8b8b0ff.

---

###### If this was not intentional, just remove the comment from your code. You can also set the [\`reopenClosed\`](https://github.com/JasonEtco/todo#configuring-for-your-project) config if you don't want this to happen at all anymore.",
}
`;
